### ISSUE TITLE
Write Vignettes

### ISSUE THREAD

**Issue by pommedeterresautee** [dmlc/xgboost#158]
I have thought to how make more people use Xgboost R package, and I think more documentation may help. 

Lots of R package have a Vignette too light (copy paste of the R functions documentation) or too complex, explaining maths behind (and finishes by my implementation is a lot better than the other 10 others existing implementation).

IMO a good Vignette should show use cases. I love the new soon to publish data.table ones. Very beautiful, I learned a lot (even if I use data.table a lot). 

So I tried to copy this example. Following is a link to a first version:
https://www.dropbox.com/s/bp8x1bmhg1bk9rf/discoverYourData.html?dl=0

I give you html directly as generating the Vignette requires using my fork of xgboost (a pull request is waiting the publising of update of another package).

The text is clearly not definitive, I want to add more technical details.
It has not as purpose to cover all the package features, just one use case among other (and probably not the one most people may expect) but it is my use case.

It may look like a blog article but I don't think it is an issue.

Please let me know what you think about this approach so I can improve the document in that way.

Kind regards,
Michaël

**Comment by tqchen** [comment]
This looks nice, and btw I think this can also be a nice material for blogpost introducing the package. 

Thanks

**Comment by pommedeterresautee** [comment]
I agree about the blog post but i have no blog with visitors where to show the article. However I plan to post a link to the vignette in some forum.

I think a vignette about prediction and in particular using cross validation, designing custom measures and so on would be nice. Something about the core use of xgboost. Unfortunately I have no real experience for this use. In general I tend to think most people don't check demo so, in some way, we should convert them to vignette (the interesting knowledge inside it).

Btw if you notice some error in the vignette, let me know.

Regards,
Michaël

**Comment by hetong007** [comment]
@pommedeterresautee actually we have a vignette file [here](https://github.com/tqchen/xgboost/tree/master/R-package/vignettes). But somehow I think it is not working anymore. I'll try to fix this issue. I suggest to write it in the official way.

I am writing an blog post introducing of `xgboost` in Chinese and I organize the features as following:

1. Basic features: `xgboost`,`predict` and `xgb.cv`
2. Speed and accuracy
3. Special features. Basically this section is listing functions with links to the correspond [demo files](https://github.com/tqchen/xgboost/tree/master/R-package/demo).

I think in our vignette, we can remove section 2. And expand the demos in section 3 into small code samples.

**Comment by pommedeterresautee** [comment]
@hetong007 I have seen the latex vignette. I have no big experience of writing vignette and may not understand the interest of the user, but in the existing vignette there is no use case described, it looks like too much to the existing R documentation.

As a new user I would expect real situation. I would ask : show me how I should drive my data analysis with xgboost. Of course it means writing stuff sometimes not part of the package itself or explaining some concepts which are supposed to be known. But that way it would have a real added value for the beginners. 

In data.table project they are about to publish 4 vignettes. The public of each of them is not the same. First one is for the beginners, and after its more and more advanced. Last one is about some optimization.

We can follow such organization. 

First vignette would present usual stuff like you are going to do in your blog post. It's a required step.

Then a second one would focus on feature importance (not complex but specific therefore more advanced). 

Third one would be much more advanced and explain use of custom measures or whatever complex xgboost can do. 

May be a last one would be about technical mathematical details (inspired from the existing slides posted in the readme). That one would be theotical.

It would asks much more work to write 4 vignettes compared to one but every potential user would find what it needs. Having one document for everything may lead to a too long document or not enough organized or missing important stuff.

Moreover bloggers would have content to copy paste for presentation of the algorithm. I wrote as a hobby some Android application in the paste and noticed how important it was to have ready to copy paste content for the articles about my applications.

Again, I am not a scientist and may miss plenty of reasons, but as a user that's what I expect from documentation.

**Comment by hetong007** [comment]
@pommedeterresautee We can indeed have more than one vignette. However we might want to keep a balance among function documentation, wiki, demo and vignette. 

For the first three,
- We have worked for a while to make our functions well documented. 
- We have a wiki page listing detailed informations for the parameters.
- We made some demos to demonstrate the usage and grammar for different functions. And even more demos on real machine learning competitions.

And we want to use vignettes to complete the missing parts. The advantage of the vignette is that the results are visible and we can add our interpretation to them. Among your suggestions, I agree on the first and partially agree on the fourth one. The reason for partially agreement is that it should be decided by @tqchen whether vignette is the best way to demonstrate these details. 

For the advanced functions, I think the difficulty lies in the uniqueness and indenpendence of the functions. 
- The uniqueness, say customized objective function, makes it harder to introduce by analogy to other packages. People need detailed guidance for grammar, result interpretation and useful scenario. 
- The functions are mostly independent to each other. I can hardly find connections between them. However we need a good way to organize them together in the vignette otherwise we are just listing them. And we have already done it in several ways.

It might be better to think about what are the features that needs introduction additional to the demos. I believe we don't have to include them all in the vignette. IMHO, customizing objective function involves some math, probably users can't understand it instantly. So it is meaningful to have detailed explanation. However, predicting with the first n trees is very straight forward and easy to learn from the demo.

Finally, the 'copy-and-paste-ability' is never a problem in R documentation. We can use RMarkdown/Knitr to render the vignette to make sure the results are reproducible from the codes.

**Comment by tqchen** [comment]
We definitely want more documentations.  I my opinion, what we are currently missing is a complete tutorial style guides.  What @pommedeterresautee  provides is indeed on of these.

Such notebook style examples allows walk user through the features by examples and we would love to have collection of them, and I think we can make them available in both github as markdown and 
vignette

**Comment by pommedeterresautee** [comment]
I am not sure to see the added value of demo code. Anything in a demo could be in a vignette or an example of function documentation.

However I fully agree with @hetong007 regarding the balance between other sources. Wiki / Blogs can be a copy paste of vignettes but between function documentation and vignette we should be careful where we put information. I have to say in the beginning I tend to put everything in documentation function but not everything have its place there. 

From my understanding, technical information should be in the documentation (in detail part) like mathematics (if it makes sense to add these information) or implementation as I am sure academics and other smart people would love to read that, but full analysis of a dataset which contains information regarding several functions should be in vignette, anything with a context. This basic rule may be a good one to respect for keeping a correct balance.

Wiki don't need to be original it's more a way to make easy to see the capabilities without installing the package. So copy paste of vignettes is cool IMO.

Regarding math and custom functions I agree. May be writing a Kaggle winning solution as a vignette (more information on the how and the why than a blog article) would be a way to illustrate the capacities of customizing measures without having to show every possibilities. In such a vignette the goal would be more to inspire than providing a ready to use method. But it may be super hard to write. It's really just an idea and may be not the most urgent thing.

Regarding the need to select functions which absolutely need to be well explained I fully agree too. It would probably be the one you want to put in your blog.

1) basic training 
2) predict 
3) improve parameters by cross validation 

Then something interesting is having several example of dataset which requires different parameters (a regression, a binary classification, a multi class classification...). Of course, I never see questions about it in the issues but linear boosting may be demonstrated  in the vignette (and a little explanation where it should be use).

Interpretation of cross validation results may require lots of lines of text too. Where each measure should be used.

Writing that I realize that with all the already implemented measures there are lots of capabilities and custom measures explanation are not so urgent :)

Kind regards,
Michaël

**Comment by hetong007** [comment]
@pommedeterresautee good points! I take the point that it is better to see the capabilities without installing the package. So we don't have to care whether there's overlap with existing stuffs.

Now we have a lot of demos. But yet they are not sufficient. To fully demonstrate the features we still need to take care all the aspect. We already have an issue on, as you mentioned, the multi class classification [here](https://github.com/tqchen/xgboost/issues/153). 

On the other hand, it is indeed difficult to show everything from a single project/data. We have many topics to introduce. Maybe we can consider the following classes of the vignette:

1. Basic functions and grammar on meaningful datasets.
2. Special features and how we can benefit under various application scenarios.
3. The mathematical details and interpretations.

So that we can have different context for our interpretations. And in each class we can have more than one vignettes. IMHO The first one is essential. We don't need to complete the latter two urgently. It might take weeks to work on them. Considering the next submission to CRAN, I personally think it is good enough to only have basic vignettes. 

I am sure more or less you know rmarkdown/knitr. I prefer this to latex style documents. If we reach to an agreement here, then we can firstly focus on the basic informations and break them into small parts recently. Then making plans and work on them could be efficient.

**Comment by pommedeterresautee** [comment]
@hetong007 this time I will write a short comment: I agree on all what you said :-)

FYI the Vignette in HTML I posted above (Dropbox link) is written in RMarkdown (code in Rmarkdown is in the pull request).

Kind regards,
Michaël

**Comment by pommedeterresautee** [comment]
This package may be useful 
https://github.com/richfitz/ghwiki/blob/master/README.md

It makes possible to create a github Wiki based on Rmarkdown documents.

**Comment by pommedeterresautee** [comment]
I have pull request a new Vignette.

Render is here:
https://www.dropbox.com/s/m8q7lkukmjukla3/xgboostPresentation.html?dl=0

And I have updated an old one:
https://www.dropbox.com/s/bp8x1bmhg1bk9rf/discoverYourData.html?dl=0

The new one is supposed to be the most important one. Today it is clearly based on the demo code. I have added some wording to make it easier to understand. It is far from complete, but it is still a basis to work on.

Let me know your opinion over them. Don't hesitate to amend them.

The CSS is the one I use at work (based on something I found on github long time ago and a lot amended since). You will enjoy the famous Deloitte netword strange blue (between violet and blue).

Kind regards,
Michael

**Comment by pommedeterresautee** [comment]
@hetong007 PDF vignette creates some temp file for the purpose of the vignette like files model.dump, xgb.Matrix and so on. They are created in `vignettes` folder.

Would it be possible to clean these file at the end of the vignette building (like adding invisible code to execute at the end of the text of the vignette)?

I have updated the code of the markdown vignettes and with devtool funciton `build_vignette()` can know update the xgboost documentation with links to the vignette. But because of the temp file, I can't push the result (of course I can exclude them from git but it would not be a clean solution).

If I can push these modification, everyone will be able to see the existing vignette without having to go on my Dropbox account, which, we can say it, is not the best solution ever!

Kind regards,
Michaël

**Comment by tqchen** [comment]
I tend not to push pdf files to github account, because they are are big and usually cause the repo to explode if we have too many updates (git have to remember all of them). 

I think the Rmarkdown can be browsed directly from the github. The vigenettes will be created and packed for CRAN version. For the github version, it would be great if we create vignette when user checkout at repo.

Tianqi

**Comment by pommedeterresautee** [comment]
I get your point and fully agree. Still PDF generation should clean temp files (I don't know Latex to do it myself).

To generate the vignette at the installation step from Github:
devtools::install_github(build_vignettes = TRUE)

Got information here:
http://r-pkgs.had.co.nz/vignettes.html

Few post above I have posted a link to a tool to generate Github pages from Vignette. Therefore we would (in theory, not tried the tool yet) have the same content in the package Vignette and in the Wiki.

Kind regards,
Michael

**Comment by pommedeterresautee** [comment]
@hetong007 Hi, I was wondering if we need the PDF Vignette. I see 4 points:
- almost all the content of the vignette is in the Rmarkdown document. I have copied paste your introduction which was much better that the one I wrote and the remaining content is a copy paste of the demo code like in the PDF ;
- to generate the PDF vignette you need to have a Latex generator installed on your computer (outside R) which can be annoying for some user ;
- it is not as easy to insert in a wiki than a HTML document ;

I will not delete any file I am not the author, therefore it is up to you to decide if you think it still make sense to keep the PDF.

In any case, I will pull request my updates in vignette writing **every day** so I am sure you and @tqchen can follow the progress and make you able to react if you don't agree with the content of the documents.

**Comment by pommedeterresautee** [comment]
@hetong007 I just add biblioagraphy to RMarkdown, in the bib file there are 3 documents, in the PDF there is only 2 citations, one doc is never used. Is it usual? Should I add a citation of the doc somewhere? It seems only cited document have complete reference at the bottom of the Rmarkdown one.

Missing document:
```Bib
@article{friedman2000additive,
  title={Additive logistic regression: a statistical view of boosting (with discussion and a rejoinder by the authors)},
  author={Friedman, Jerome and Hastie, Trevor and Tibshirani, Robert and others},
  journal={The annals of statistics},
  volume={28},
  number={2},
  pages={337--407},
  year={2000},
  publisher={Institute of Mathematical Statistics}
}

```

**Comment by pommedeterresautee** [comment]
I plan to write a third Vignette based on advanced demo code (may be latter after the Cran update publishing): 
- predict from leaf
- predict first trees
- predict leaf indice
- cross validation process (not in demo code).

All these parts deal with improving / understanding trees in a model. It makes more sense to put them together.

Regarding the existing vignette, I still need to add some text about linear learning but I don't plan to do big changes on text content or part organization / presentation / CSS. 

The only thing I will try to improve is the articulation between parts. Right now, they are a bit too much independent. I will try to make it look like more to a smooth flow.

Please @tqchen @hetong007 let me know your opinion on these documents as they look like now. I don't want to start polishing them, reading and rereading documents if you don't like the existing result. 

Kind regards,
Michael

**Comment by hetong007** [comment]
Hi Michael,

Thanks for the efforts. I'll check the missing reference thing. 

Because of [this issue](https://github.com/tqchen/xgboost/issues/170), I studied about vignettes a bit more. And I have some questions and hope we can have a discussion.

Firstly, it turns out that to let `R CMD build` automatically deal with our `Rmd` vignettes, we need to add several lines at the beginning of the file:
```
<!--
%\VignetteEngine{knitr}
%\VignetteIndexEntry{Discover Your Data}
-->
```
[Here](https://raw.githubusercontent.com/yihui/knitr/7eb34104d5410645f5c0fa35b4f7906e2bcc1cb8/inst/doc/knitr-markdown.Rmd)'s an official example. The side-effect is that the YAML marks are not working anymore. For a possible fix for the customized css, we could take a look at the bottom of [this page](http://cran.r-project.org/web/packages/knitr/vignettes/knitr-html.html). 

So natually, if we use `R CMD build` to generate vignettes, then we no longer need to push pdf/html to this github repository.  Things will be perfect if:

1. We use `R CMD build` to automatically generate all the vignette(which is of course preferred).
2. We can solve the formatting, e.g. css, issues for `Rmd` files.
3. We can solve [issue 170](https://github.com/tqchen/xgboost/issues/170).

Please share any experiences and thoughts. Thanks!

**Comment by pommedeterresautee** [comment]
Hi @hetong007 

First thank you a lot for taking a look to the Vignette.

Regarding the bug you refer to, I think I already fixed it in the version of the vignette included in the pull request.

Basically, as you can see in my fork [here](https://raw.githubusercontent.com/pommedeterresautee/xgboost/master/R-package/vignettes/discoverYourData.Rmd) the lines you make a reference to are now included in the YAML:
```
---
title: "Understand your dataset with Xgboost"
output: 
  rmarkdown::html_vignette:
    css: vignette.css
    number_sections: yes
    toc: yes
author: Tianqi Chen, Tong He, Michaël Benesty
vignette: >
  %\VignetteIndexEntry{Discover your data}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
```

According to my experience, Vignette is correctly built with devtools and included in R Xgboost documentation (with links and so on) and I keep my CSS.

Can you let me know which command is supposed to work?

When I execute make Rpack in Ubuntu I get:
```
...
R CMD build xgboost
* checking for file ‘xgboost/DESCRIPTION’ ... OK
* preparing ‘xgboost’:
* checking DESCRIPTION meta-information ... OK
* cleaning src
* installing the package to build vignettes
* creating vignettes ... OK
* cleaning src
* excluding invalid files
...
* checking compiled code ... OK
* checking sizes of PDF files under ‘inst/doc’ ... OK
* checking installed files from ‘inst/doc’ ... OK
* checking files in ‘vignettes’ ... OK
* checking examples ... OK
...
```

As you can see, Rpack seems to create Vignette correctly. And YAML works (I see it in the generated Vignette).

Let me know if there is something I miss.

**Comment by pommedeterresautee** [comment]
@hetong007 regarding the PDF:
- in the pull request I have finally fixed the temp doc issue (I added hidden code to clean generated files like dump models...) ;
- I thought a lot about the need to have a PDF with the Rmarkdown and finally I think it has advantages to have both format at the same time. Rmarkdown can be like a tutorial (but much more complete than a classical blog entry) and the PDF can provide a more academic approach, meaning everything is inside, but it is dense and short. In some way each format is adapted to its public, that is the advantage I see to have both documents in parallel.

My only remaining concern is about the need to have Latex to generate the documentation. For external users wanting to install the Github version we will need to explain both how to do to avoid documentation generation (if they don t want to install Latex) and provide information on how to get this documentation (install Latex and so on for each platform).

Let me know your thoughts.

Kind regards,
Michael

**Comment by hetong007** [comment]
@pommedeterresautee Thanks for the solution to the YAML issue.

Do you mind to check the `NOTE`s from the `R CMD check` command? There's one thing keep haunting me for a while:
```
...
* checking if this is a source package ... NOTE
Found the following apparent object files/libraries:
  src/src/gbm/gbm.o src/src/io/io.o src/src/tree/updater.o
  src/subtree/rabit/src/engine_empty.o src/wrapper/xgboost_wrapper.o
Object files/libraries should not be included in a source package.
...
```
I can reproduce it on several machines. Can you check if it appears on yours as well? 

If we run `R CMD build --no-build-vignettes xgboost` then there's no pre-built `.o` file anymore. But usually building vignettes will not affect other files. It could be either our mistakes, or a bug in `R CMD build`.

I tend to try harder on our side first. If we can't figure out the truth, another less elegant solution is manually delete those 5 files.

**Comment by pommedeterresautee** [comment]
I have the same note with build Rpack.

Just to be sure, when you test, you use the RPack part commands in Makefile, right?

**Comment by hetong007** [comment]
@pommedeterresautee Yes, I think right now one can reproduce it with simply

```
git clone https://github.com/tqchen/xgboost && cd xgboost && make Rpack
```

I am asking it on [stackoverflow](http://stackoverflow.com/questions/28521074/r-cmd-build-leaves-compiled-o-files-in-src-folder). Hoping somebody could help us.

**Comment by pommedeterresautee** [comment]
Regarding Vignette content, I just read an interesting article:
http://www.win-vector.com/blog/2015/02/does-balancing-classes-improve-classifier-performance/comment-page-1/#comment-59429

It shows poor performance for RF in rare event use case. I have not tried it but it seems logic to me that an algo working on sub sampled data performed badly on rare events.

It makes me wonder if it is a good idea to have some algo comparison in the Vignettes on some specific case. In importance feature one I put a word regarding drawback of RF for this specific use case.

Not related, I present Xgboost to a friend who is - unlike me - a real data scientist. He uses Scipy at work. When I told him that I read tons of times that Xgboost was much more rapid than Scipy, the first thing he told me was: show me benchmarks. I don't know if it should be in a Vignette or just the Wiki / README but I think we would need that too.

Kind regards,
Michael

**Comment by tqchen** [comment]
I think benchmark can be in the demo part, by creating a benchmark folder. There are some existing ones that serves as benchmarks

* https://github.com/tqchen/xgboost/tree/master/demo/kaggle-higgs
* https://github.com/tqchen/xgboost/tree/master/demo/yearpredMSD

The higgs boson contains speedtest against sklearn. We can create another issue on this

**Comment by pommedeterresautee** [comment]
A package I was not aware of. I would be interesting to include it in the benchmark:
http://blog.revolutionanalytics.com/2015/03/a-first-look-at-rxbtrees.html

FYI Revolution 7.2 is freely available for people doing Kaggle competition.

Kind regards,
Michael
